CHANGELOG.md

```diff
@@ -0,0 +1,167 @@
+- `sapling_crypto::constants` module.
+- `sapling_crypto::keys`:
+  - `SpendAuthorizingKey`
+  - `SpendValidatingKey`
+- `sapling_crypto::note_encryption`:
+  - `CompactOutputDescription` (moved from
+    `zcash_primitives::transaction::components::sapling`).
+  - `SaplingDomain::new`
+  - `Zip212Enforcement`
+- `sapling_crypto::prover::{SpendProver, OutputProver}`
+- `sapling_crypto::tree::Node::{from_bytes, to_bytes}`
+- `sapling_crypto::value`:
+  - `ValueCommitTrapdoor::from_bytes`
+  - `impl Sub<TrapdoorSum> for TrapdoorSum`
+  - `impl Sub<CommitmentSum> for CommitmentSum`
+- `sapling_crypto::zip32` module (moved from `zcash_primitives::zip32::sapling`).
+- `impl Debug for sapling_crypto::keys::{ExpandedSpendingKey, ProofGenerationKey}`
+- Test helpers, behind the `test-dependencies` feature flag:
+  - `sapling_crypto::prover::mock::{MockSpendProver, MockOutputProver}`
+
+### Changed
+- `sapling_crypto`:
+  - `BatchValidator::validate` now takes the `SpendVerifyingKey` and
+    `OutputVerifyingKey` newtypes.
+  - `SaplingVerificationContext::new` now always creates a context with ZIP 216
+    rules enforced, and no longer has a boolean for configuring this.
+  - `SaplingVerificationContext::{check_spend, final_check}` now use the
+    `redjubjub` crate types for `rk`, `spend_auth_sig`, and `binding_sig`.
+  - `SaplingVerificationContext::{check_spend, check_output}` now take
+    the `PreparedSpendVerifyingKey` and `PreparedOutputVerifyingKey`
+    newtypes.
+  - `SaplingVerificationContext::final_check` now takes its `value_balance`
+    argument as `V: Into<i64>` instead of `Amount`.
+- `sapling_crypto::address::PaymentAddress::create_note` now takes its `value`
+  argument as a `NoteValue` instead of as a bare `u64`.
+- `sapling_crypto::builder`:
+  - `SaplingBuilder` no longer has a `P: consensus::Parameters` type parameter.
+  - `SaplingBuilder::new` now takes a `Zip212Enforcement` argument instead of a
+    `P: consensus::Parameters` argument and a target height.
+  - `SaplingBuilder::add_spend` now takes `extsk` by reference. Also, it no
+    longer takes a `diversifier` argument as the diversifier may be obtained
+    from the note.
+  - `SaplingBuilder::add_output` now takes an `Option<[u8; 512]>` memo instead
+    of a `MemoBytes`.
+  - `SaplingBuilder::build` no longer takes a prover, proving context, progress
+    notifier, or target height. Instead, it has `SpendProver, OutputProver`
+    generic parameters and returns `(UnauthorizedBundle, SaplingMetadata)`. The
+    caller can then use `Bundle::<InProgress<Unproven, _>>::create_proofs` to
+    create spend and output proofs for the bundle.
+  - `Error` has new error variants:
+    - `Error::DuplicateSignature`
+    - `Error::InvalidExternalSignature`
+    - `Error::MissingSignatures`
+- `sapling_crypto::bundle`:
+  - `Bundle` now has a second generic parameter `V`.
+  - `Bundle::value_balance` now returns `&V` instead of `&Amount`.
```

Comment:
Should references like `Amount` refer back to the types in `zcash_primitives` more explicitly? Someone just looking at this changelog otherwise doesn't have a good way to distinguish where each type can be found.

Reply:
The types that were moved around inside `zcash_primitives::sapling` before extraction were logically in `sapling_crypto` so I renamed those as such. For the others, I left explicit `zcash_primitives` paths where they already existed, so yeah we should also update the implicitly-pathed ones that are now external.